### PR TITLE
[WIP] shield/blocking rework

### PR DESCRIPTION
**Pull request by unid15**
- Swords only block melee attacks now. Eswords still block bullets. All shields block both melee attacks and projectiles
- Facing your attacker increases your chance to block his attacks by flat 30%. Facing away from your attacker decreases your chance to block his attacks by flat 30%. 

The chance modifiers go like this: +30% (from the front) / +10% / +-0% (from the side) / -10% / -30% (from behind)
- **List of items that can be used for blocking + their efficiency**
- Shields (100%)
  - Bucklers (33.3%)
- Dual eswords (100%)
- HF Katanas, eswords (80%)
- Claymores, katanas (70%)
- Forged swords (50%)
- The final block chance (after direction modifier) is multiplied by that percentage
- Also swords and shields (no eswords unfortunately) now play a sound when blocking an attack

### CODE REVIEW

**Comment by Imagebard** [comment]
1.Did you edit all weapons? Katanas(/obj/item/weapon/katana) have IsShield checks too. Costume Dispenser shields also have IsShield rolls, but they're children of Riot Shields last time I checked so It should be fine.

**Comment by AutismSpider** [comment]
Costume shields are children of riot so it should be fine.

He needs look at E Shield katanas and the high frequency blade (that uses Esword blocking i think)

**Comment by unid15** [comment]
Yeah I changed all instances of isShield()

**Comment by Imagebard** [comment]
I edited my post 5 times.
I am done editing it now.
Can you read the final result of it?

**Comment by AutismSpider** [comment]
@Imagebard By dual do you mean holding two ESwords or the Double-bladed ESword?

**Comment by unid15** [comment]
> Also, is there a way to make certain weapons have a +% chance of blocking something? Dual e-sword should have bigger block rate than a normal e-sword.

I think dual eswords already have 2x the block rate (because their offhands count as shields, too). I'll look into making block % customizable

**Comment by Imagebard** [comment]
Double-bladed ESword.

**Comment by 9600bauds** [comment]
This kind of makes eswords completely shit over tasers if you just keep running towards the dude

**Comment by Imagebard** [comment]
You're gonna have to look for alternative ways of blocking it.
Bolas or a stun baton should work.

**Comment by AutismSpider** [comment]
@9600bauds  Can do that right now with the use of that vest and riot shield Esword on top of all this forces you to use a shot gun with stun rounds/beanbag or bolas spam to deal with that

**Comment by 9600bauds** [comment]
I mean with JUST the esword.
I'm not sure but I think stun rounds wouldn't work either

**Comment by Imagebard** [comment]
@clusterfack It doesn't have conflicts anymore

**Comment by Brolaire** [comment]
what happened to this?

**Comment by Imagebard** [comment]
COOOOONFLIIIIIIICTS

File: code/game/objects/items.dm [hunk 1]
```diff
@@ -35,6 +35,20 @@
 	var/cant_drop = 0 //If 1, can't drop it from hands!
 
 	var/armor = list(melee = 0, bullet = 0, laser = 0,energy = 0, bomb = 0, bio = 0, rad = 0)
+	var/list/properties = list()
```
Comment (ComicIronic) [inline code/game/objects/items.dm:38]: No list on all items. Add null checks, remove list().

File: code/game/objects/items.dm [hunk 2]
```diff
@@ -699,10 +744,36 @@
...
+					base_block_chance += 30
+				if(45) //From diagonal - +10%
+					base_block_chance += 10
+				//if(90) //From side - nothing
+				if(135) //From behind, diagonal - -10%
+					base_block_chance -= 10
+				if(180) //From behind - -30%
+					base_block_chance -= 30
+
+			base_block_chance = Clamp(base_block_chance, 0, 80) //If you're facing away from the attack, the chance to block it may become 0%.
```
Comment (ComicIronic) [inline code/game/objects/items.dm:767]: Why 80 and also magic number 80

File: code/game/objects/items.dm [hunk 3]
```diff
@@ -699,10 +744,36 @@
...
+
+		if(properties["blocking"])
```
Comment (ComicIronic) [inline code/game/objects/items.dm:769]: Yeah, so if you want that on an item level then I suggest adding a GetProperty(string) proc for this.
Also make that a precompile def, you shouldn't have strings if they need to be repeated consistently.

File: code/game/objects/items/weapons/twohanded.dm [hunk 4]
```diff
@@ -48,6 +48,8 @@
 /obj/item/offhand/proc/attach_to(var/obj/item/I)
 	I.wielded = src
 	wielding = I
+	properties = I.properties.Copy()
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/twohanded.dm:51]: Make this a proc as well, GetAllProperties() gets list or null.

**Comment by ComicIronic** [comment]
RIP this until it gets a rework, because the properties idea is decent.

**Pull request closed without merging**